backend/src/org/commcare/session/CommCareSession.java

```diff
@@ -248,15 +253,30 @@ public SessionDatum getNeededDatum() {
      * @return A session datum definition if one is pending. Null otherwise.
      */
     public SessionDatum getNeededDatum(Entry entry) {
-        int nextVal = getData().size();
-        //If we've already retrieved all data needed, return null.
-        if (nextVal >= entry.getSessionDataReqs().size()) {
+        Vector<SessionDatum> allDatumsNeeded = entry.getSessionDataReqs();
+        OrderedHashtable datumsSoFar = getCollectedDatums();
+
+        if (datumsSoFar.size() >= allDatumsNeeded.size()) {
```

Comment:
This actually isn't a sufficient check, because the frame can contain additional data that is irrelevant (is kind of a long story).

either way, would prefer to stick with doing the full check.